assignment2/q2.cpp

```diff
@@ -0,0 +1,148 @@
+#include <iostream>
+using namespace std;
+
+struct Node{
+    int value;
+    Node *parent;
+    Node *left, *right;
+};
+
+bool FindNode(Node *root, Node *to_find) {
+
+    if(root == NULL)
+        return false;
+
+    if(root == to_find)
+        return true;
+
+    if(FindNode(root->left, to_find) || FindNode(root->right, to_find))
+        return true;
+
+    return false;
+}
+
+//O(N^2) without any additional memory or changing structure
+//could be O(N) if i.e. it stores paths to nodes and then compares them
+Node *FindCommonAncestor(Node *root, Node *first, Node *second) {
+
+    if(root == NULL)
+        return NULL;
+
+    //nodes are not in the tree
+    if(!FindNode(root, first) || !FindNode(root, second))
+        return NULL;
+
+    //if both are in the left tree
+    if(FindNode(root->left, first) && FindNode(root->left, second))
+        return FindCommonAncestor(root->left, first, second);
+
+    //if both are in the right tree
+    if(FindNode(root->right, first) && FindNode(root->right, second))
+        return FindCommonAncestor(root->right, first, second);
+
+    //if nodes are one in left, other in right tree
+    else
```

Comment:
No need for this `else`, since there is a `return` in the `if` before it.
You did rely on the early returns in the previous cases of this function, no reason not to rely on it here too.